src/cpp/fastdds/xtypes/utils.cpp

```diff
@@ -0,0 +1,1031 @@
+}
+
+ReturnCode_t json_serialize_member(
+        const traits<DynamicDataImpl>::ref_type& data,
+        MemberId member_id,
+        TypeKind member_kind,
+        const std::string& member_name,
+        nlohmann::json& output,
+        DynamicDataJsonFormat format) noexcept
+{
+    switch (member_kind)
+    {
+        case TK_NONE:
+        case TK_BOOLEAN:
+        case TK_BYTE:
+        case TK_INT8:
+        case TK_INT16:
+        case TK_INT32:
+        case TK_INT64:
+        case TK_UINT8:
+        case TK_UINT16:
+        case TK_UINT32:
+        case TK_UINT64:
+        case TK_FLOAT32:
+        case TK_FLOAT64:
+        case TK_FLOAT128:
+        case TK_CHAR8:
+        case TK_CHAR16:
+        case TK_STRING8:
+        case TK_STRING16:
+        case TK_ENUM:
+        case TK_BITMASK:
+        {
+            return json_serialize_basic_member(data, member_id, member_kind, member_name, output, format);
+        }
+        case TK_STRUCTURE:
+        case TK_BITSET:
+        {
+            traits<DynamicDataImpl>::ref_type st_data =
+                    traits<DynamicData>::narrow<DynamicDataImpl>(data->loan_value(member_id));
+            if (nullptr != st_data)
+            {
+                ReturnCode_t ret = RETCODE_OK;
+                nlohmann::json j_struct;
+                DynamicTypeMembersById members;
+                if (RETCODE_OK != (ret = st_data->enclosing_type()->get_all_members(members)))
+                {
+                    EPROSIMA_LOG_WARNING(XTYPES_UTILS,
+                            "Error encountered while serializing structure/bitset member to JSON: get_all_members failed.");
+                }
+                else
+                {
+                    for (auto it : members)
+                    {
+                        if (RETCODE_OK != (ret = json_serialize_member(st_data, it.second, j_struct, format)))
+                        {
+                            EPROSIMA_LOG_WARNING(XTYPES_UTILS,
+                                    "Error encountered while serializing structure/bitset member's member to JSON.");
+                            break;
+                        }
+                    }
+                }
+                if (RETCODE_OK == ret)
+                {
+                    json_insert(member_name, j_struct, output);
+                }
+                // Return loaned value
+                ReturnCode_t ret_return_loan;
+                if (RETCODE_OK != (ret_return_loan = data->return_loaned_value(st_data)))
+                {
+                    EPROSIMA_LOG_WARNING(XTYPES_UTILS, "Error encountered while returning loaned value.");
+                }
+                // Give priority to prior error if occurred
+                if (RETCODE_OK != ret)
+                {
+                    return ret;
+                }
+                else
+                {
+                    return ret_return_loan;
+                }
```

Comment:
Any nicer way of doing this?

Reply:
Changed last part for a ternary operator, I found it an overkill to encapsulate this in a function.